File: app/src/main/java/com/albineli/udacity/popularmovies/model/MovieListStateModel.java
package com.albineli.udacity.popularmovies.model;

import android.os.Bundle;
import android.support.annotation.Nullable;

import com.albineli.udacity.popularmovies.enums.MovieListFilterDescriptor;

import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;

public class MovieListStateModel {
    public final int pageIndex;
    public final @MovieListFilterDescriptor.MovieListFilter int filter;
    public final List<MovieModel> movieList;
    public final int selectedMovieIndex;
    public final int firstVisibleMovieIndex;

    private MovieListStateModel(List<MovieModel> movieList, @MovieListFilterDescriptor.MovieListFilter int filter, int pageIndex, int selectedMovieIndex, int firstVisibleMovieIndex) {
        this.movieList = movieList;
        this.filter = filter;
        this.pageIndex = pageIndex;
        this.selectedMovieIndex = selectedMovieIndex;
        this.firstVisibleMovieIndex = firstVisibleMovieIndex;
    }

    public static void saveToBundle(Bundle bundle, List<MovieModel> movieList, int filter, int pageIndex, int selectedMovieIndex, int firstVisibleMovieIndex) {
        bundle.putParcelableArrayList(MOVIE_LIST_BUNDLE_KEY, new ArrayList<>(movieList));
        bundle.putInt(PAGE_INDEX_BUNDLE_KEY, pageIndex);
        bundle.putInt(SELECTED_MOVIE_INDEX_BUNDLE_KEY, selectedMovieIndex);
        bundle.putInt(FIRST_VISIBLE_MOVIE_INDEX_BUNDLE_KEY, firstVisibleMovieIndex);
        bundle.putInt(FILTER_BUNDLE_KEY, filter);
    }

    public static @Nullable MovieListStateModel getFromBundle(Bundle bundle) {
        if (bundle == null) {
            return null;
        }

        @MovieListFilterDescriptor.MovieListFilter int filter = bundle.getInt(FILTER_BUNDLE_KEY);
        int pageIndex = bundle.getInt(PAGE_INDEX_BUNDLE_KEY, Integer.MIN_VALUE);
        int selectedMovieIndex = bundle.getInt(SELECTED_MOVIE_INDEX_BUNDLE_KEY, Integer.MIN_VALUE);
        int firstVisibleMovieIndex = bundle.getInt(FIRST_VISIBLE_MOVIE_INDEX_BUNDLE_KEY, Integer.MIN_VALUE);
        ArrayList<MovieModel> movieList = bundle.getParcelableArrayList(MOVIE_LIST_BUNDLE_KEY);

        return new MovieListStateModel(movieList, filter, pageIndex, selectedMovieIndex, firstVisibleMovieIndex);
    }

    public static @NotNull MovieListStateModel getFromArguments(Bundle arguments) {
        int filter = MovieListFilterDescriptor.parseFromInt(arguments.getInt(FILTER_BUNDLE_KEY));
        return new MovieListStateModel(null, filter, Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MIN_VALUE);
    }

    private static String SELECTED_MOVIE_INDEX_BUNDLE_KEY = "selected_movie_index_bundle_key";
    private static String FILTER_BUNDLE_KEY = "filter_bundle_key";
    private static String PAGE_INDEX_BUNDLE_KEY = "page_index_bundle_key";
    private static String MOVIE_LIST_BUNDLE_KEY = "movie_list_bundle_key";
    private static String FIRST_VISIBLE_MOVIE_INDEX_BUNDLE_KEY = "first_visible_movie_index_bundle_key";
}

File: app/src/main/java/com/albineli/udacity/popularmovies/model/MovieListStateModel.kt
package com.albineli.udacity.popularmovies.model

import android.os.Bundle

import com.albineli.udacity.popularmovies.enums.MovieListFilterDescriptor

import java.util.ArrayList

data class MovieListStateModel constructor(val movieList: List<MovieModel>, @MovieListFilterDescriptor.MovieListFilter val filter: Int, val pageIndex: Int, val selectedMovieIndex: Int, val firstVisibleMovieIndex: Int) {

    companion object {
        fun saveToBundle(bundle: Bundle, movieList: List<MovieModel>, filter: Int, pageIndex: Int, selectedMovieIndex: Int, firstVisibleMovieIndex: Int) {
            bundle.putParcelableArrayList(MOVIE_LIST_BUNDLE_KEY, ArrayList(movieList))
            bundle.putInt(PAGE_INDEX_BUNDLE_KEY, pageIndex)
            bundle.putInt(SELECTED_MOVIE_INDEX_BUNDLE_KEY, selectedMovieIndex)
            bundle.putInt(FIRST_VISIBLE_MOVIE_INDEX_BUNDLE_KEY, firstVisibleMovieIndex)
            bundle.putInt(FILTER_BUNDLE_KEY, filter)
        }

        fun getFromBundle(bundle: Bundle?): MovieListStateModel? {
            if (bundle == null) {
                return null
            }

            @MovieListFilterDescriptor.MovieListFilter val filter = bundle.getInt(FILTER_BUNDLE_KEY)
            val pageIndex = bundle.getInt(PAGE_INDEX_BUNDLE_KEY, Integer.MIN_VALUE)
            val selectedMovieIndex = bundle.getInt(SELECTED_MOVIE_INDEX_BUNDLE_KEY, Integer.MIN_VALUE)
            val firstVisibleMovieIndex = bundle.getInt(FIRST_VISIBLE_MOVIE_INDEX_BUNDLE_KEY, Integer.MIN_VALUE)
            val movieList = bundle.getParcelableArrayList<MovieModel>(MOVIE_LIST_BUNDLE_KEY)

            return MovieListStateModel(movieList, filter, pageIndex, selectedMovieIndex, firstVisibleMovieIndex)
        }

        fun getFromArguments(arguments: Bundle): MovieListStateModel {
            val filter = MovieListFilterDescriptor.parseFromInt(arguments.getInt(FILTER_BUNDLE_KEY))
            return MovieListStateModel(emptyList(), filter, Integer.MIN_VALUE, Integer.MIN_VALUE, Integer.MIN_VALUE)
        }

        private val SELECTED_MOVIE_INDEX_BUNDLE_KEY = "selected_movie_index_bundle_key"
        private val FILTER_BUNDLE_KEY = "filter_bundle_key"
        private val PAGE_INDEX_BUNDLE_KEY = "page_index_bundle_key"
        private val MOVIE_LIST_BUNDLE_KEY = "movie_list_bundle_key"
        private val FIRST_VISIBLE_MOVIE_INDEX_BUNDLE_KEY = "first_visible_movie_index_bundle_key"
    }
}
